### ISSUE TITLE
Guaranteed MIT build

### ISSUE THREAD

**Issue by makaimann** [msoos/cryptominisat#510]
Currently it appears that M4RI is included by default if it’s installed. For CVC4 we have a --bsd flag which disables all GPL components. Would you consider adding a -mit compilation flag (as opposed to just -DNOM4RI) to guarantee that the executable is under the MIT license regardless of what’s installed on the system. This can be important for users/companies that need MIT but are working on a shared system where M4RI could be installed somewhere. It’s also forward-looking in the sense that -mit could continue to be supported even if other optional non-MIT compliant plugins are added.

Thanks!

**Comment by msoos** [comment]
I'm sorry but this makes no sense. First of all, "-mit" is not a flag that I would implement because it has to be "--mit", right? I wouldn't design "-mit", with a single dash, it just isn't good practice. Secondly, I cannot change "cmake" -- I hope you realise that adding "--mit" would imply changing a humongous system that is written by hundreds of professionals just so we can have our own flag? Anyway, of course that's not possible, I will not edit, package, test and maintain a multi-million line program to have my own flag. That just makes no sense, I hope you understand that.

What we COULD have is "-DMIT=ON" -- that's the cmake way of doing this. However, we already have "-DNOM4RI=ON". I can implement "-DMIT=ON" but that would just be an an alias. However, I would have to spend time and effort to build, test and maintain this flag forever. Are you sure that's worth it? Why isn't  "-DNOM4RI=ON" sufficient?

Conclusion is: please use "-DNOM4RI=ON" or please give me a meaningful, reasoned text about why I would have to re-build, package and maintain a multi-million line build system for the sake of a company that cannot pass "-DNOM4RI=ON" and instead wants single-dash multi-letter options which have been deprecated for decades. I'm closing this because the request is simply not meaningful.

**Comment by sschnug** [comment]
Hey Mate, thanks for all your hard work on this project.

I have to admit, that your reaction above sounds a bit harsh (at least to me).
I thought this was a reasonable request (and i'm not even comparing to the average github issue standard of well-used projects), which probably also was more focused on the *what* and less on the *how* (of course nobody changes cmake to get some stylized flags).

I did not spend much time on checking all that license-stuff and therefore i try to explain my feelings here with my limited knowledge (and i'm ready to be punished for my ill-preparation):

Let's focus on `-DMIT=ON` vs. `DNOM4RI=ON`.

- The current master/Readme.md contains no license-information / hints except for `LICENSE.txt`
  - This one boils down to: `see the copyright headers` which effects in the user checking all the sources
- I think i read somewhere about M4RI vs. MIT, but this might be hidden somewhere in the code, the issue-tracker or your blog

The core issue imho is, and i think i would interpret the original post in this way, using `DNOM4RI=ON` to get something MIT-compatible needs additional knowledge about the inner state of CM. Is M4RI the only thing i need to care with my system or not? Maybe this new version added additional non-MIT code (linked to OPs: *It’s also forward-looking in the sense that -mit could continue to be supported even if other optional non-MIT compliant plugins are added.*)

While every company using this project should spend some time in checking the code and it's licenses *themself* (not blindly trusting `-DMIT=ON`), this feature would help with prototyping code when evaluating if CM can help (e.g. does CM help with an MIT-configuration; does it help without XORs).

I understand, that this effects in maintenance costs, but the developer himself has the best view on all those components and their licenses and could help the user here.

To me, these  *license-based compilation-flags* sound great and maybe it's something approved by the users, comparing to other projects with huge commercial usage like ffmpeg (and it's `--enable-gpl` and `--enable-nonfree` flags + an additional [License FAQ](https://www.ffmpeg.org/legal.html)). Yes... i know... it's a huge project with much support.

Thanks for reading!

**Comment by makaimann** [comment]
Yes, my apologies on the -mit vs. --mit. I meant --mit. And you're both right about that, I don't actually care about the syntax at all. I must admit I don't know anything about cmake, and really just meant something like -DMIT=ON.

**Comment by msoos** [comment]
Hey, my apologies, I got a bit frustrated -- what happens lots of time is that people want a small feature and they are happy when I give it to them but the maintenance of features over a period of 7 years can be a lot more expensive than imagined. Anyway, I am implementing this feature now, this issue will get closed when I push the feature in :) Thanks for being patient and standing your ground :)

**Comment by makaimann** [comment]
No worries -- I certainly understand that. Thanks for implementing it!

You've already implemented this so it's a bit late, but I feel that I should comment on the reasoning a bit further as well. @sschnug already did a great job summarizing (and you're completely correct by the way) so I won't reiterate those points too much. This feature is not really for me or the people I work with, because thanks to your guidance we already know to use -NOM4RI=ON to guarantee MIT. The main idea is that if the goal is to get more companies to use CM (which I understand isn't necessarily the goal), it just gives the technical employees something to point the lawyers to. Of course, the company will have to understand all the dependencies and licenses, but it should help ease the initial evaluation and demonstrate that the license is "clean". In other words, it is and should be an alias for other options. More generally, my personal opinion is that it should be very easy (without understanding the codebase) to compile with the correct license regardless of the environment you're building in. Either by defaulting or providing a simple command line option. 

Anyway, I hope this isn't too much of a maintenance hassle for you and thanks for the all the great work on this project!

**Comment by msoos** [comment]
Although I implemented the flag, let me speak my mind here, as I disagree with the above:

1) I think the LICENSE.txt is pretty clear. I strongly disagree with the statement "This one boils down to: see the copyright headers which effects in the user checking all the source files". This is not true. The LICENSE.txt states that everything is MIT that is needed for compilation. Nobody needs to go around looking at individual file headers unless you want to verify that statement. I have now added a line to this stating that linking in M4RI can make it GPL.

2) -DNOM4RI=ON always guaranteed an MIT build. There was no need for a flag that is an alias to NOM4RI. They do the same thing.

The flag is now there. I will now pay the price of maintaining it -- nobody else will, as you have clearly demonstrated that you are unwilling to create a pull request to fulfil your own need. Instead, I am must do it, for free, for a company that clearly wants to close-source their product hence make money out of free work of others but cannot spend time to implement the feature they need. Instead, they seem to argue endlessly, with walls of text -- interesting approach. I hope they don't pay their employees with arguments spanning pages instead of real money.

I consider this request closed. In case you need any other feature, please create a pull request.
